lib/path.cpp

```diff
@@ -245,13 +251,101 @@ bool Path::isHeader(const std::string &path)
     return startsWith(extension, ".h");
 }
 
-Standards::Language Path::identify(const std::string &path, bool *header)
+static bool hasEmacsCppMarker(const char* path)
+{
+    // TODO: identify is called three times for each file
+    // Preprocessor::loadFiles() -> createDUI()
+    // Preprocessor::preprocess() -> createDUI()
+    // TokenList::createTokens() -> TokenList::determineCppC()
+#ifdef LOG_EMACS_MARKER
+    std::cout << path << '\n';
+#endif
+
+    FILE *fp = fopen(path, "rt");
+    if (!fp)
+        return false;
+    std::string buf(128, '\0');
+    {
+        // TODO: read the whole first line only
+        const char * const res = fgets(const_cast<char*>(buf.data()), buf.size(), fp);
+        fclose(fp);
+        fp = nullptr;
+        if (!res)
+            return false; // failed to read file
+    }
+    // TODO: replace with regular expression
+    const auto pos1 = buf.find("-*-");
+    if (pos1 == std::string::npos)
+        return false; // no start marker
+    const auto pos_nl = buf.find_first_of("\r\n");
+    if (pos_nl != std::string::npos && (pos_nl < pos1)) {
+#ifdef LOG_EMACS_MARKER
+        std::cout << path << " - Emacs marker not on the first line" << '\n';
+#endif
+        return false; // not on first line
+    }
+    const auto pos2 = buf.find("-*-", pos1 + 3);
+    // TODO: make sure we have read the whole line before bailing out
+    if (pos2 == std::string::npos) {
+#ifdef LOG_EMACS_MARKER
+        std::cout << path << " - Emacs marker not terminated" << '\n';
+#endif
+        return false; // no end marker
+    }
+#ifdef LOG_EMACS_MARKER
+    std::cout << "Emacs marker: '"  << buf.substr(pos1, (pos2 + 3) - pos1) << "'" << '\n';
+#endif
+    // TODO: support /* */ comments
+    const std::string buf_trim = trim(buf); // trim whitespaces
+    if (buf_trim[0] != '/' || buf_trim[1] != '/') {
+#ifdef LOG_EMACS_MARKER
+        std::cout << path << " - Emacs marker not in a comment: '"  << buf.substr(pos1, (pos2 + 3) - pos1) << "'" << '\n';
+#endif
+        return false; // not a comment
+    }
+
+    // there are more variations with lowercase and no whitespaces
+    // -*- C++ -*-
+    // -*- Mode: C++; -*-
+    // -*- Mode: C++; c-basic-offset: 8 -*-
+    std::string marker = trim(buf.substr(pos1 + 3, pos2 - pos1 - 3), " ;");
+    // cut off additional attributes
+    const auto pos_semi = marker.find(';');
+    if (pos_semi != std::string::npos)
+        marker.resize(pos_semi);
+    findAndReplace(marker, "mode:", "");
+    findAndReplace(marker, "Mode:", "");
+    marker = trim(marker);
+    if (marker == "C++" || marker == "c++") {
+        // NOLINTNEXTLINE(readability-simplify-boolean-expr) - TODO: FP
```

Comment:
Upstream issue: https://github.com/llvm/llvm-project/issues/93157.

This is not the actual issue but related.